File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
import sys

input = sys.stdin.readline

from collections import deque

dx1 = [-1, 0, 1, 0]
dy1 = [0, 1, 0, -1]

dx2 = [-1, -1, -2, -2, 1, 1, 2, 2]
dy2 = [-2, 2, -1, 1, 2, -2, 1, -1]

dx3 = [-1, 0, 1, 0, -1, -1, -2, -2, 1, 1, 2, 2]
dy3 = [0, 1, 0, -1, -2, 2, -1, 1, 2, -2, 1, -1]
k = int(input())

m, n = map(int, input().split())

arr = [list(map(int, input().split())) for _ in range(n)]

visited = [[[0] * (k + 1) for _ in range(m)] for _ in range(n)]

q = deque()

q.append((0, 0, 0))
visited[0][0][0] = 1
flag = 1
while q:
    x, y, z = q.popleft()
    if x == n - 1 and y == m - 1:
        flag = 0
        print(visited[x][y][z] - 1)
        break
    if z < k:
        for i in range(12):
            nx = x + dx3[i]
            ny = y + dy3[i]

            if 0 <= nx < n and 0 <= ny < m:
                if arr[nx][ny] == 0:
                    if i <= 3 and visited[nx][ny][z] == 0:
                        q.append((nx, ny, z))
                        visited[nx][ny][z] = visited[x][y][z] + 1
                    elif i > 3 and visited[nx][ny][z+1] == 0:
                        q.append((nx, ny, z + 1))
                        visited[nx][ny][z + 1] = visited[x][y][z] + 1
    else:
        for i in range(4):
            nx = x + dx1[i]
            ny = y + dy1[i]
            if 0 <= nx < n and 0 <= ny < m:
                if arr[nx][ny] == 0 and visited[nx][ny][z] == 0:
                    q.append((nx, ny, z))
                    visited[nx][ny][z] = visited[x][y][z] + 1

if flag:
    print(-1)

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.js
const fs = require("fs");
const input =
    process.platform === "linux"
        ? fs.readFileSync("/dev/stdin").toString().split("\n")
        : fs.readFileSync("example.txt").toString().split("\r\n");

const dx1 = [-1, 0, 1, 0]
const dy1 = [0, 1, 0, -1]
const dx2 = [-1, 0, 1, 0, -1, -1, -2, -2, 1, 1, 2, 2]
const dy2 = [0, 1, 0, -1, -2, 2, -1, 1, 2, -2, 1, -1]


const k = parseInt(input[0])
const [m, n] = input[1].split(" ").map((e) => parseInt(e))

const arr = new Array()

for (let i = 2; i < n + 2; i++) {
    const a = input[i].split(" ").map((e) => parseInt(e));
    arr.push(a)
}

const visited = Array.from(Array(n), () => Array.from(Array(m), () => Array(k + 1).fill(0)))

const q = [];

q.push([0, 0, 0])
visited[0][0][0] = 1

let flag = 1;

while (q.length > 0) {
    const [x, y, z] = q.shift();

    if (x == n - 1 && y == m - 1) {
        flag = 0
        console.log(visited[x][y][z] - 1)
        break
    }

    if (z < k) {
        for (let i = 0; i < 12; i++) {
            let nx = x + dx2[i]
            let ny = y + dy2[i]

            if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                if (arr[nx][ny] == 0) {
                    if (i <= 3 && visited[nx][ny][z] == 0) {
                        q.push([nx, ny, z])
                        visited[nx][ny][z] = visited[x][y][z] + 1
                    } else if (i > 3 && visited[nx][ny][z + 1] == 0) {
                        q.push([nx, ny, z + 1])
                        visited[nx][ny][z + 1] = visited[x][y][z] + 1

                    }
                }
            }
        }
    } else {
        for (let i = 0; i < 4; i++) {
            let nx = x + dx1[i]
            let ny = y + dy1[i]
            if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                if (arr[nx][ny] == 0 && visited[nx][ny][z] == 0) {
                   q.push([nx, ny, z])
                    visited[nx][ny][z] = visited[x][y][z] + 1
                }
            }
        }
    }
}

if (flag) {
    console.log(-1)
}